client/js/analysis/gdpr.js

```diff
@@ -67,14 +79,117 @@ export async function getStreamingData(data) {
             Promise
                 .all(contentPromises)
                 .then((contentWithIndices) => {
-                  const content = contentWithIndices
-                      .map((entry) => entry.data)
-                      .reduce((dataA, dataB) => dataA.concat(dataB), []);
-                  resolve(content);
+                  resolve(
+                      contentWithIndices
+                          .map((entry) => entry.data)
+                          .reduce((dataA, dataB) => dataA.concat(dataB), [])
+                          .map((entry) => {
+                            // dates are strings in JSON, convert to JS date
+                            const endTime = new Date(entry.endTime);
+
+                            // move according to UTC offset
+                            const endTimeMs = +endTime;
+                            const endTimeOffset =
+                              endTime.getTimezoneOffset() * 60000;
+
+                            entry.endTime = new Date(endTimeMs - endTimeOffset);
+                            return entry;
+                          }),
+                  );
                 });
           });
         },
         (error) => reject(error),
     );
   });
 }
+
+/** @typedef {Map<string, Map<string, number>>} ArtistTrackTimeMap */
+/** @typedef {{
+ * start: Date,
+ * totals: ArtistTrackTimeMap
+ * }} ArtistTrackTimeWindow */
+
+/**
+ * Calculates aggregates from a list of GDPR records.
+ *
+ * @param {GDPRRecord[]} data The list of GDPR records to aggregate.
+ * @returns {{totals: ArtistTrackTimeMap, windows: ArtistTrackTimeWindow[]}} An
+ * object with two properties: `totals` and `windows`. `totals` is a map whose
+ * keys are artists and whose values are maps from track names to duration
+ * listened. `windows` is an array of window objects. Each window object has a
+ * `start` representing the start of that window, and a `totals`, representing
+ * the durations listened to each song during that window.
+ */
+export function collateStreamingData(data) {
+  let index = 0;
+  let currentRecord = data[0];
+
+  const WINDOW_SIZE = 86400000;
+  // round down to the start of the day of the first record
+  let windowStart =
+    +currentRecord.endTime - (+currentRecord.endTime % WINDOW_SIZE);
+
+  const windows = [];
+
+  // a map of maps: key is artist name, value is a map where key
+  // is track name and value is time listened to that song in total
+  /** @type {ArtistTrackTimeMap} */
+  const totals = new Map();
+
+  // each window is a map of maps: key is artist name, value is a map where key
+  // is track name and value is time listened to that song during this window
+  /** @type {ArtistTrackTimeMap} */
+  let window = new Map();
```

Comment:
"window" is a bit confusing of a name, because here it's an ArtistTrackTimeMap but "windows" is an array of ArtistTrackTimeWindow's (what I'm assuming to be an actual "window"). Consider renaming to something like "windowTotals" to make the while loop below more readable.

Reply:
I subbed out the term "window" for "interval".